commit a99f0ee43403b172acb6101f7ad30777d359bd61
Author:     Yan Cheng <58191769+yanchengnv@users.noreply.github.com>
AuthorDate: Wed Sep 13 14:59:42 2023 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Wed Sep 13 14:59:42 2023 -0400

    Support big job workspace (#1988)
    
    * develop binary download
    
    * support binary download to admin client
    
    * fix formatting
    
    * fix type hints
    
    * fix imports
    
    * remove Union from hint
    
    * consolidate consts, added chunk tests
    
    * do not create workspace symlink before job done
    
    ---------
    
    Co-authored-by: Isaac Yang <isaacy@nvidia.com>
---
 nvflare/private/fed/server/job_runner.py | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/nvflare/private/fed/server/job_runner.py b/nvflare/private/fed/server/job_runner.py
index 8dc0d578..0d733db2 100644
--- a/nvflare/private/fed/server/job_runner.py
+++ b/nvflare/private/fed/server/job_runner.py
@@ -15,6 +15,7 @@
 import json
 import os
 import shutil
+import tempfile
 import threading
 import time
 from typing import Dict, List, Tuple
@@ -28,7 +29,7 @@ from nvflare.apis.job_def import ALL_SITES, Job, JobMetaKey, RunStatus
 from nvflare.apis.job_scheduler_spec import DispatchInfo
 from nvflare.apis.workspace import Workspace
 from nvflare.fuel.utils.argument_utils import parse_vars
-from nvflare.fuel.utils.zip_utils import zip_directory_to_bytes
+from nvflare.fuel.utils.zip_utils import zip_directory_to_file
 from nvflare.lighter.utils import verify_folder_signature
 from nvflare.private.admin_defs import Message, MsgHeader, ReturnCode
 from nvflare.private.defs import RequestHeader, TrainingTopic
@@ -382,11 +383,13 @@ class JobRunner(FLComponent):
         job_id = fl_ctx.get_prop(FLContextKey.CURRENT_JOB_ID)
         workspace = Workspace(root_dir=self.workspace_root)
         run_dir = workspace.get_run_dir(job_id)
-        workspace_data = zip_directory_to_bytes(run_dir, "")
         engine = fl_ctx.get_engine()
         job_manager = engine.get_component(SystemComponents.JOB_MANAGER)
-
-        job_manager.save_workspace(job_id, workspace_data, fl_ctx)
+        with tempfile.TemporaryDirectory() as td:
+            output_file = os.path.join(td, "workspace")
+            zip_directory_to_file(run_dir, "", output_file)
+            job_manager.save_workspace(job_id, output_file, fl_ctx)
+            self.log_info(fl_ctx, f"Workspace zipped to {output_file}")
         shutil.rmtree(run_dir)
 
     def run(self, fl_ctx: FLContext):
